### PR TITLE
fix(mdx): share module and bundle resolution across concurrent renders

### PR DESCRIPTION
**Pull request by kwakayama**
Fixes veryfront/veryfront-issue-inbox#1510
Fixes veryfront/veryfront-issue-inbox#1511

## Cause

On 2026-09-19 03:40:30 UTC, 10 concurrent `GET /` requests for one hosted project reached a production pod about 100s after it started. None completed. One request's trace made 1,044 calls to `/projects/:id/cache/*`: 998 individual `/cache/get` calls (p50 1.0s, 312 hit the 10s timeout), 985 of them under `mdx.fetch_module < utils.parallelMap < mdx.process_vf_modules`. The `api-cache-http` circuit breaker opened at 03:41:02, about 2,000 fast failures followed, the MDX transform tree timed out at 47-66s (#1510), and the SSR pipeline hit its 60s deadline (#1511). The same pattern appeared on 2026-09-15 01:16 on another pod (2,076 request timeouts). Operator memory recycling (since 2026-09-12) replaces 11-26 production pods per day, so cold pods are now common.

Two multipliers on a cold pod:

1. `fetchAndCacheModule` deduplicates in-flight modules only within one `processVfModuleImports` call (`context.inFlightModules`). N concurrent cold renders of the same page each walk the whole `_vf_modules` graph and repeat every distributed cache read.
2. Bundle recovery dominates the reads. Log analysis of a single cold production render (2,080 `/cache/get` in one request) shows each `_vf_modules` section recovering a graph of about 247 HTTP bundles. `ensureHttpBundlesExist` fetches the code in batches (`get-batch`), but then looks up each bundle's recovery identity with its own `get` calls (identity + import map, about 2 per bundle, which matches the ~517 gets per section in the trace). Sibling sections recovered the same bundle graph at the same time.

## Fix

- **Process-wide single flight for entry module fetches** (`module-fetcher/shared-module-fetches.ts`). Entry fetches (no parent module) with the same identity share one in-flight resolution. The key covers project ID, content source, ESM cache directory, project directory, local-project flag, compile mode, React version, dependency snapshot key, module server origin, server external packages, missing-module mode, and entry path. Results never cross projects or content versions.
  - Only entry fetches are shared. Nested fetches stay scoped to the resolution that owns them, and a call made from inside a shared resolution never joins another one, so two resolutions cannot wait on each other.
  - Entries are removed when they settle. Resolved paths are then served from the module path cache, and a rejection is shared with current waiters only. The next request retries.
  - `invalidateModulePaths` and `clearModulePathCache` reset the map, so a request that starts after a content change never joins a resolution that read the old source.
  - Joined renders still record the resolved modules into their own render session (route module manifest).
  - A never-settling resolution stops accepting new joiners after 60s (the existing `Singleflight` stale guard).
  - Per-request limits still apply to joined renders. Modules the shared resolution recorded count toward the joined render's 500-module graph limit. If the shared resolution fails with the leading render's `TransformTreeTimeoutError`, a joined render retries alone within its own deadline.
- **Missing HTTP bundles are fetched once per cache directory across callers** (`bundle-recovery.ts`). `ensureHttpBundlesExist` claims missing hashes synchronously before its first await. Other callers wait for the claim, then read the result from disk, and recover any bundle the claimant could not write. A claim is held until the bundle is written or single-bundle recovery settles. A caller releases all of its claims before it waits on other claims. Recovery that runs under a held claim is marked with AsyncLocalStorage. Nested `ensureHttpBundlesExist` calls inside it fetch in-flight bundles themselves instead of waiting, so claim holders never wait on each other.
- **Batched recovery identity lookups** (`HttpBundleCache.getBatchRecoveryIdentities`). Identity records, shared import maps (read once per fingerprint), and original URLs are read with `getBatch` instead of one `get` per bundle. The semantics match `getIdentityMetadata` + `getOriginalUrl`.

The 30s transform-tree deadline semantics are unchanged.

## Tests (red before the fix)

- `module-fetcher/index.test.ts`, "process-wide module fetch single-flight": 10 concurrent cold resolutions of one entry graph (page -> a, b -> c) against a distributed cache stub that counts `get` and adds 200ms latency. Before: 80 cache gets and 40 source reads (10x). After: 8 cache gets (same as one solo cold graph) and 4 source reads. Also covers a joined render retrying alone after the leading render's deadline, graph-limit admission for joined renders, cross-project isolation of the same path, a shared rejection that is retried by the next request (before: 3 source reads for 3 concurrent callers, and they did not all reject), no joining across an invalidation, and session attribution for every joined render.
- `shared-module-fetches.test.ts` (hermetic): key identity per input, single run per key, rejection not retained, synchronous failure, nested-call bypass, reset, and session replay.
- `bundle-recovery.test.ts`: 30 bundles recovered with zero single-key reads and one import-map read (before: 60 single reads). 5 concurrent `ensureHttpBundlesExist` calls fetch each of 20 bundles once (before: 100 code reads, after: 20). A bundle a concurrent claimant failed to fetch is retried by the waiter. Single-bundle recovery for 3 concurrent callers runs once (before: 3 direct code reads).

Local: `deno task lint:ci`, `deno task typecheck`, `deno fmt --check`, and `deno task test:file` for `src/transforms/`, `src/modules/react-loader/`, `src/modules/manifest/`, `src/rendering/`, `src/cache/`, and `src/server/context/` all pass.

## Staging verification

Pending. This PR is updated after the rc build rolls out to staging.

## Summary by CodeRabbit

* **Performance**
  * Improved concurrent loading of HTTP bundles and MDX modules by sharing in-flight work and reducing duplicate fetches.
  * Added batched bundle recovery to reduce cache requests.

* **Reliability**
  * Improved recovery when bundle or module fetches are incomplete or time out, including safe retries.
  * Improved consistency after cache invalidation by clearing stale shared fetch state.

* **Caching**
  * Ensured concurrently rendered module graphs receive the modules resolved by shared fetches.

### CODE REVIEW

**Comment by kwakayama** [comment]
## Code Review: 84/100 — Good, minor suggestions

Well-engineered concurrency fix backed by real production trace data, with unusually thorough concurrency reasoning and test coverage; main gaps are the still-pending staging verification and some subtle timing edge cases worth double-checking before merge.

**Strengths**
- The root-cause writeup (pod trace analysis, 03:40 UTC incident, ~2,080 `/cache/get` calls per render) is excellent — it justifies the fix with hard evidence rather than intuition, and the fix maps 1:1 onto the two multipliers identified (per-render dedup only vs. process-wide, and per-bundle identity lookups vs. batched).
- Concurrency safety is handled carefully: entry-only sharing (`isEntryFetch = parentModulePath === undefined && lineage.size === 0`) plus the nested-call bypass in `runSharedModuleFetch` (`if (sharedResolutionScope.getStore()) return await resolve();`) prevents two shared resolutions from ever waiting on each other — a real deadlock risk this design correctly avoids.
- Cache key correctness: `getSharedModuleFetchKey` folds in every input that affects the resolved module (project, content source, cache/project dirs, compile mode, React version, dependency snapshot, module server origin, external packages, missing-module mode, entry path), and it's unit-tested field-by-field against accidental cross-project/version sharing.
- Test coverage is genuinely strong: `shared-module-fetches.test.ts` covers key stability, rejection-not-retained, sync-throw, nested bypass, and reset; `index.test.ts` adds an end-to-end 10-concurrent-cold-render scenario with call counting; `bundle-recovery.test.ts` covers the batched-identity path and a caller that must retry after a concurrent claimant's fetch fails. The `session replay` test specifically verifies modules resolved by the leader get recorded into every joined render's own manifest session.
- `resetSharedModuleFetches()` on `invalidateModulePaths`/`clearModulePathCache` correctly stops new joiners from reading stale content without disturbing in-flight waiters already attached to the old resolution.

**Concerns / suggestions**
1. **Staging verification is explicitly "Pending"** in the PR description. This PR exists specifically to fix a live production incident (two recent timeouts), and it replaces the hot path for *every* MDX render with new global mutable state (`bundleFetchesInFlight`, `sharedModuleFetches`). Given the blast radius, I'd want at least one staging soak under realistic concurrent cold-pod load before merging to main, not just local `deno task test:file`.
2. In `fetchMissingBundles`'s `recoverFromMiss` path, the claim is released *before* `recoverHttpBundleByHash` actually completes (`claims?.release(hash)` then `await recoverHttpBundleByHash(...)`). This is intentional (avoids recursive self-deadlock) and is covered by the "retries a bundle that a concurrent caller failed to fetch" test, but it means a waiting caller's claim wait can resolve well before the bundle is actually on disk, so the "single flight" benefit silently degrades to "every waiter retries independently" for this fallback path. Worth a one-line comment in the code noting this is a deliberate correctness-over-throughput tradeoff (the existing comment explains the deadlock avoidance but not the throughput implication) so a future reader doesn't "fix" it into a deadlock.
3. `getSharedModuleFetchKey` includes `context.serverExternalPackages ?? []` directly in `JSON.stringify`; if that array's element order isn't guaranteed stable across two otherwise-identical contexts, functionally-equivalent renders would fail to join the same resolution (a missed-optimization, not a correctness bug, but cheap to guard against — e.g. sort before stringifying).
4. Two new pieces of process-wide mutable state (`bundleFetchesInFlight` map, `sharedModuleFetches` singleflight instance) with fairly intricate claim/release/replay timing. This is justified by the incident, but it raises the bar for anyone touching this code later — consider whether `getBundleFetchesInFlightCount`/`getSharedModuleFetchCount` should also be surfaced as metrics (not just test hooks) so a similar incident is visible sooner next time, rather than only discoverable via a 47-66s timeout trace.
5. Automated reviews (CodeRabbit, Codex) were still pending/rate-limited at review time — worth checking their output once available, particularly Codex's security pass, before merge.

Given the quality of the analysis and tests, my main ask before merge is item 1 (staging confirmation) and a quick look at item 2/3 — nothing here blocks review, but I'd treat "staging verification pending" as a real gate given this is fixing a production SSR outage.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/transforms/mdx/esm-module-loader/module-fetcher/index.ts [hunk 1]
```diff
@@ -300,6 +302,19 @@ export async function fetchAndCacheModule(
...
     )
+    : isEntryFetch
+    // Concurrent requests for the same entry share one resolution of its graph.
+    ? runSharedModuleFetch(
+      getSharedModuleFetchKey(context, bindingKey),
+      () =>
+        doFetchAndCacheModule(
+          normalizedPath,
+          context,
+          fetchAndCacheModuleFn,
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/transforms/mdx/esm-module-loader/module-fetcher/index.ts]: Fixed in 44af84e: a joined render retries alone when the shared resolution fails with TransformTreeTimeoutError from the leading render's deadline (test: "retries alone when the leading render hits its own deadline").

File: src/transforms/mdx/esm-module-loader/module-fetcher/shared-module-fetches.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            path: await resolve(),
+            recordedModules,
+          })),
+      );
+    },
+    { staleAfterMs: SHARED_MODULE_FETCH_STALE_AFTER_MS },
+  );
+
+  for (const modulePath of result.recordedModules) recordModuleToSession(modulePath);
+  return result.path;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/transforms/mdx/esm-module-loader/module-fetcher/shared-module-fetches.ts:198]: Fixed in 44af84e: modules recorded by the shared resolution are admitted into each joined render's moduleGraph, which raises ModuleGraphLimitError for that render only (test: "counts shared modules toward each joined render's graph limit").

File: src/transforms/esm/bundle-recovery.ts [hunk 3]
```diff
@@ -307,6 +307,196 @@ export async function recoverHttpBundleByHash(
...
+    for (const m of missing) context.onFailed(m.hash);
+    return;
+  }
+
+  const codes = await httpBundleCache.getBatchCodes(missing.map((m) => m.hash));
+  const identities = await httpBundleCache.getBatchRecoveryIdentities([...codes.keys()]);
+
+  const recoverFromMiss = async (hash: string, canonicalPath: string): Promise<void> => {
+    claims?.release(hash);
+    const recovered = await recoverHttpBundleByHash(
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/transforms/esm/bundle-recovery.ts]: Fixed in 44af84e: single-bundle recovery now keeps its claim until it settles. Recovery that runs under a held claim is marked with AsyncLocalStorage, and nested ensureHttpBundlesExist calls inside it fetch in-flight bundles themselves instead of waiting, so claim holders never wait on each other (test: "keeps single-bundle recovery claimed so concurrent callers do not repeat it").

**Comment by kwakayama** [comment]
@codex review

File: src/transforms/esm/bundle-recovery.ts [hunk 4]
```diff
@@ -374,94 +597,54 @@ export async function ensureHttpBundlesExist(
...
+    await Promise.all(claims.inFlight.values());
+    const leftover: MissingBundle[] = [];
+    for (const entry of missing) {
+      if (!claims.inFlight.has(entry.hash)) continue;
+      const bundle = await readCachedHttpBundleFile(fs, entry.canonicalPath);
+      if (bundle && !isDegradedArtifact(bundle.code)) fetchContext.onMaterialized(bundle.code);
+      else leftover.push(entry);
+    }
+    // The other caller could not materialize these; recover them here.
+    await fetchMissingBundles(leftover, fetchContext);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/transforms/esm/bundle-recovery.ts]: Fixed in e40b358: a waiter whose claim holder left the bundle missing claims the retry itself (MAX_BUNDLE_CLAIM_ROUNDS rounds of claim, wait, re-check) before any unclaimed last-resort fetch. Test: "claims a retry owner when a concurrent caller leaves a bundle missing" asserts 2 code reads for 3 concurrent callers.

File: src/transforms/mdx/esm-module-loader/module-fetcher/index.ts [hunk 5]
```diff
@@ -292,21 +304,40 @@ export async function fetchAndCacheModule(
...
+    // Concurrent requests for the same entry share one resolution of its graph.
+    fetchPromise = runSharedModuleFetch(
+      getSharedModuleFetchKey(context, bindingKey),
+      resolveModule,
+      {
+        // Modules another request resolved still count toward this request's
+        // graph limit.
+        onResolved: (recordedModules) => admitSharedModules(moduleGraph, recordedModules),
+        // The leading request's deadline is not this request's deadline.
+        retryAloneOn: (error) => error instanceof TransformTreeTimeoutError,
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/transforms/mdx/esm-module-loader/module-fetcher/index.ts]: Fixed in e40b358: retryAloneOn now also covers ModuleGraphLimitError, so a render that joined a resolution which exhausted the leading render's graph resolves with its own. Test: "retries alone when the leading render has no room in its graph".

File: src/transforms/mdx/esm-module-loader/module-fetcher/index.ts [hunk 6]
```diff
@@ -292,21 +304,40 @@ export async function fetchAndCacheModule(
...
     );
+  } else if (isEntryFetch) {
+    // Concurrent requests for the same entry share one resolution of its graph.
+    fetchPromise = runSharedModuleFetch(
+      getSharedModuleFetchKey(context, bindingKey),
+      resolveModule,
+      {
+        // Modules another request resolved still count toward this request's
+        // graph limit.
+        onResolved: (recordedModules) => admitSharedModules(moduleGraph, recordedModules),
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/transforms/mdx/esm-module-loader/module-fetcher/index.ts]: Fixed in e40b358: the shared resolution now records the paths it admits to its module graph (recordSharedModuleAdmission, set at admission time rather than from manifest records), and joined renders admit exactly those. Test: "admits modules a shared resolution could not resolve into joined graphs" covers a stubbed dependency in lenient mode.

**Comment by kwakayama** [comment]
@codex review

File: src/transforms/esm/bundle-recovery.ts [hunk 7]
```diff
@@ -374,94 +603,64 @@ export async function ensureHttpBundlesExist(
...
-            cacheHttpModule,
-            undefined,
-            fallbackIdentity,
-          );
-          if (!recovered) failed.add(hash);
-          return;
-        }
+      if (waitedFor.length === 0) {
+        outstanding = [];
+        break;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/transforms/esm/bundle-recovery.ts:640]: Fixed in 02e8c99: before reporting failures, ensureHttpBundlesExist re-checks each failed hash on disk and, for one a concurrent caller materialized, re-runs the dependency walk for it (one non-recursing pass) and drops it from the failure set only when it and its dependencies are present. The concurrent retry test now asserts that at most one caller reports the bundle, since whether the owner finishes before the retry owner writes it is a real race; a new test covers a bundle written by another caller mid-fetch.

**Comment by kwakayama** [comment]
@codex review

File: src/transforms/esm/bundle-recovery.ts [hunk 8]
```diff
@@ -374,94 +605,92 @@ export async function ensureHttpBundlesExist(
...
+        await ensureHttpBundlesExist(
+          materialized,
+          cacheDir,
+          cacheHttpModule,
+          fallbackIdentity,
+          false,
+        ),
+      );
+      for (const { hash } of materialized) {
+        if (!stillFailed.has(hash)) failed.delete(hash);
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/transforms/esm/bundle-recovery.ts:695]: Fixed in 85e3164e05. `stillFailed` is merged into `failed` before the materialized roots are dropped, so a dependency the recursive pass could not recover survives the recheck. Regression test: `reports a dep missing from a bundle another caller materialized` in `src/transforms/esm/bundle-recovery.test.ts` — a concurrent caller materializes `A` while `A` imports an unavailable `B`, and `ensureHttpBundlesExist` now returns `["B"]` instead of `[]`.

File: src/transforms/esm/bundle-recovery.ts [hunk 9]
```diff
@@ -374,94 +605,92 @@ export async function ensureHttpBundlesExist(
...
+      }
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/transforms/esm/bundle-recovery.ts:696]: Fixed in 85e3164e05, taking your committable suggestion as written (`for (const hash of stillFailed) failed.add(hash);` ahead of the materialized-delete loop). Every hash in `stillFailed` comes from the dependency closure of a materialized bundle, so it is genuinely unrecoverable and genuinely needed. Regression test: `reports a dep missing from a bundle another caller materialized` in `src/transforms/esm/bundle-recovery.test.ts`, which fails on the previous code and passes now.

File: src/transforms/esm/bundle-recovery.test.ts [hunk 10]
```diff
@@ -327,6 +388,214 @@ describe("transforms/esm/bundle-recovery", () => {
...
+    it("stops reporting a bundle another caller materialized", async () => {
+      const cacheDir = await makeTempDir();
+      const hash = "904";
+      const code = "export const written = true;\n";
+      const backend = createCountingBatchBackend({}, { latencyMs: 20 });
+      __setDistributedCacheAccessorForTests(() => Promise.resolve(backend));
+
+      try {
+        const bundles = [{ path: join(cacheDir, `http-${hash}.mjs`), hash }];
+        const pending = ensureHttpBundlesExist(bundles, cacheDir, () => Promise.resolve(null));
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/transforms/esm/bundle-recovery.test.ts:598]: Fixed in 85e3164e05. `createCountingBatchBackend` now exposes a `fetchStarted` promise that resolves on the first backend read, and `stops reporting a bundle another caller materialized` awaits it before `writeTextFile`, so the post-fetch materialized-failure recheck is the path under test rather than the trivial initial local hit. The new `reports a dep missing from a bundle another caller materialized` test uses the same checkpoint.

File: src/transforms/mdx/esm-module-loader/module-fetcher/shared-module-fetches.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+      },
+      { staleAfterMs: SHARED_MODULE_FETCH_STALE_AFTER_MS },
+    );
+  } catch (error) {
+    if (leading || !options.retryAloneOn?.(error)) throw error;
+    return await resolve();
+  }
+
+  options.onResolved?.(result.admittedModules);
+  for (const modulePath of result.recordedModules) recordModuleToSession(modulePath);
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/transforms/mdx/esm-module-loader/module-fetcher/shared-module-fetches.ts:197]: Fixed in 9ce657f519. Confirmed reachable: `processVfModuleImports` resolves several `_vf_modules` entries concurrently through `parallelMap` in one request sharing `context.inFlightModules`, so when entry E1 reaches dep D first and E2 takes the `return existingPromise` path, D and its whole subtree were recorded only under E1s `runWithModuleRecorder` scope. The same aliasing hit `admittedModules`: `recordSharedModuleAdmission` runs before the in-flight check so D itself was admitted for E2, but D-s subtree was admitted only under E1s `admittedModuleScope`, under-counting the graph-limit replay to renders joining E2.

The fix covers both: `moduleRecorderStorage` and `admittedModuleScope` now hold a stack and append rather than replace, each fetch collects what it records and admits into a `BorrowedModules` record keyed by the promise siblings join through (a `WeakMap`, so `ModuleFetcherContext` is untouched and the record dies with the promise), and the join path replays that record into the joiner-s scopes. Regression test: `records modules borrowed from a sibling entry into every joined session` in `module-fetcher/index.test.ts` — two entries of one render both import `c`, two further renders join one entry each, and both the replayed route-module manifest and the joined `moduleGraph` now contain `c`. It fails on the previous code.

One note for @kwakayama: this is a design choice on your hot path. Making the recorder a stack and attaching a record to each in-flight fetch is the smallest complete fix I can see, and it follows your earlier replay fixes in 44af84e/e40b358, but you may prefer to keep manifest recording outside the shared resolution entirely. The behaviour to reach is not in doubt; the mechanism is yours to redirect.

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kojiwakayama** [comment]
## Review pass — all four threads addressed, CI green

Two additive commits on `fix/process-wide-module-singleflight` (no rebase, no force-push, no amend).

**85e3164e05 — `fix(esm): propagate dependency failures from the materialized recheck`**
- Closes the codex P2 and the CodeRabbit Major thread, which are the same defect. The materialized-failure recheck dropped the recursive result: a materialized parent was cleared from `failed` whenever `stillFailed` lacked it, discarding the dependency hashes the recursive pass reported, so a bundle a concurrent caller wrote whose own dep is unavailable was reported as recovered while importing it still failed. Took CodeRabbit-s committable suggestion verbatim.
- Closes the CodeRabbit Minor test-robustness thread: `createCountingBatchBackend` now exposes a `fetchStarted` promise resolving on the first backend read, and the materialized-bundle tests await it before `writeTextFile`, so the post-fetch recheck is the path under test rather than the trivial initial local hit.
- New test `reports a dep missing from a bundle another caller materialized`; verified it fails on the previous code.

**9ce657f519 — `fix(mdx): record dependencies borrowed from a sibling entry fetch`**
- Closes the codex P2 on `shared-module-fetches.ts:145`. When two entries of one render resolve concurrently through `parallelMap` and both need the same dep, the one taking the `return existingPromise` path in `fetchAndCacheModule` recorded nothing, so the dep and its subtree lived only in the other entry-s recorder scope and a render joining only the borrowing entry replayed a partial route-module manifest. The same aliasing under-counted `admittedModules` for the graph-limit replay.
- `moduleRecorderStorage` and `admittedModuleScope` now hold a stack and append instead of replacing; each fetch collects what it records and admits into a `BorrowedModules` record keyed by the promise siblings join through (a `WeakMap`, so `ModuleFetcherContext` is untouched and the record dies with the promise); the join path replays it into the joiner-s scopes.
- New test `records modules borrowed from a sibling entry into every joined session`; verified it fails on the previous code.

**Declined:** nothing — all four findings checked out against the code.

**Gates run locally from the repo-pinned tooling:** `deno check --no-lock` on the four changed source files, `deno task lint:ci` (full chain, including `lint:testing-front-door`, `lint:test-typecheck` and `docs:api-reference:check` — no docs regeneration needed), `deno fmt --check`, and `deno task test:file` over `src/transforms/esm/` (40 files, 950 steps) and `module-fetcher/` (20 files, 299 steps). All green.

**Status:** every check run is success or skipped, the combined commit status is success, the `Automated review` gate went green on this head (the earlier red was purely the 30-minute timeout firing about two minutes before the review landed), and all 12 review threads are resolved. `mergeable_state` is now `clean`.

**Left for you, @kwakayama:**
1. The PR body-s `## Staging verification — Pending` is still yours to complete. Nothing here substitutes for it, and #1510/#1511 are production incidents whose fix is worth confirming on a cold staging pod.
2. The second commit is a design choice on your hot path. Turning the module recorder into a stack and attaching a record to each in-flight fetch is the smallest complete fix I could see and it follows your earlier replay work in 44af84e/e40b358, but you may prefer to keep manifest recording outside the shared resolution entirely. The behaviour to reach is not in doubt; the mechanism is yours to redirect. Noted in the thread too.
